### PR TITLE
Refactor ee code xin

### PR DESCRIPTION
**Pull request by jxchinaren**

### CODE REVIEW

File: src/ee/storage/TupleBlock.h [hunk 1]
```diff
@@ -35,6 +35,7 @@
 #include <deque>
 
 namespace voltdb {
+#define INVALID_NEW_BUCKET_INDEX -1
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: #defines are for C hackers. Use "static const int".

File: src/ee/storage/TupleBlock.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@ TupleBlock::TupleBlock(Table *table, TBBucketPtr bucket) :
         m_activeTuples(0),
         m_nextFreeTuple(0),
         m_lastCompactionOffset(0),
-        m_tuplesPerBlockDivNumBuckets(m_tuplesPerBlock / static_cast<double>(TUPLE_BLOCK_NUM_BUCKETS)),
+        m_tuplesPerBucket(m_tuplesPerBlock / static_cast<double>(TUPLE_BLOCK_NUM_BUCKETS)),
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.cpp]: This member is not actually needed at all -- neither is double math.

File: src/ee/storage/TupleBlock.h [hunk 3]
```diff
@@ -114,37 +115,35 @@ class TupleBlock {
         return false;
     }
 
+    /**
+     * If the next free tuple is not able to be in the current bucket, return the new bucket index,
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: It's not the "next free tuple" in the buckets, it's the "tuple block with its current fullness".

File: src/ee/storage/TupleBlock.h [hunk 4]
```diff
@@ -114,37 +115,35 @@ class TupleBlock {
...
-        }
-        else {
-            int index = static_cast<int>(::floor(m_activeTuples / m_tuplesPerBlockDivNumBuckets));
-            assert(index < TUPLE_BLOCK_NUM_BUCKETS);
-            assert(index >= 0);
-            return index;
+            return INVALID_NEW_BUCKET_INDEX;
         }
+
+        int index = static_cast<int>(::floor(m_activeTuples / m_tuplesPerBucket));
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: Better just to use integer math:
int index = TUPLE_BLOCK_NUM_BUCKETS * m_activeTuples / m_tuplesPerBlock;

File: src/ee/storage/TupleBlock.h [hunk 5]
```diff
@@ -114,37 +115,35 @@ class TupleBlock {
...
+        assert(index < TUPLE_BLOCK_NUM_BUCKETS);
+        assert(index >= 0);
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: We don't need to assert that multiplying and dividing non-negative numbers doesn't give a negative number.

File: src/ee/storage/TupleBlock.h [hunk 6]
```diff
@@ -153,6 +152,9 @@ class TupleBlock {
 
     std::pair<int, int> merge(Table *table, TBPtr source, TupleMovementListener *listener = NULL);
 
+    /**
+     * Find next free tuple storage address and its bucket index position.
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: not really "its bucket index position", more like "the tupleblock's bucket index".

File: src/ee/storage/TupleBlock.h [hunk 7]
```diff
@@ -167,12 +169,14 @@ class TupleBlock {
         }
         m_activeTuples++;
         int newBucketIndex = calculateBucketIndex();
-        if (newBucketIndex != m_bucketIndex) {
-            m_bucketIndex = newBucketIndex;
-            return std::pair<char*, int>(retval, newBucketIndex);
-        } else {
-            return std::pair<char*, int>(retval, -1);
+        if (newBucketIndex == m_bucketIndex) {
+            // have space in current bucket
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: It's not about "having space in the bucket", it's about "tupleblock is not too full for its current bucket".

File: src/ee/storage/persistenttable.cpp [hunk 8]
```diff
@@ -230,7 +230,7 @@ void PersistentTable::nextFreeTuple(TableTuple *tuple) {
     /**
      * Check to see if the block needs to move to a new bucket
      */
-    if (retval.second != -1) {
+    if (retval.second != INVALID_NEW_BUCKET_INDEX) {
```
Comment (paulmartel) [inline src/ee/storage/persistenttable.cpp]: You missed a bunch more magic "-1"s still in this module.

File: src/ee/storage/persistenttable.cpp [hunk 9]
```diff
@@ -1422,15 +1423,15 @@ bool PersistentTable::doCompactionWithinSubset(TBBucketMap *bucketMap) {
...
         for (int ii = 0; ii < TUPLE_BLOCK_NUM_BUCKETS; ii++) {
-            lightestIterator = (*bucketMap)[ii]->begin();
-            if (lightestIterator != (*bucketMap)[ii]->end()) {
+            lightestIterator = (*bucketVector)[ii]->begin();
+            if (lightestIterator != (*bucketVector)[ii]->end()) {
                 lightest = lightestIterator.key();
                 if (lightest != fullest) {
                     foundLightest = true;
                     break;
                 } else {
```
Comment (paulmartel) [inline src/ee/storage/persistenttable.cpp]: Don't need an else after a break;

File: src/ee/storage/TupleBlock.h [hunk 10]
```diff
@@ -167,12 +169,14 @@ class TupleBlock {
...
+            return std::pair<char*, int>(retval, INVALID_NEW_BUCKET_INDEX);
         }
+
+        // needs a new bucket and update bucket index
+        m_bucketIndex = newBucketIndex;
+        return std::pair<char*, int>(retval, newBucketIndex);
```
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: Even simpler?:

    if (newBucketIndex == m_bucketIndex) {
        newBucketIndex = INVALID_NEW_BUCKET_INDEX;
    }
    else {
         m_bucketIndex = newBucketIndex;
    }
    return std::pair<char*, int>(retval, newBucketIndex);

File: src/ee/storage/TupleBlock.h (hunk 1 above)
Comment (paulmartel) [inline src/ee/storage/TupleBlock.h]: "INVALID" may be a little misleading. A lot of times it just means that the bucket index has not changed, so there is not a new value.  maybe call this "NO_NEW_BUCKET_INDEX" instead?

**Comment by paulmartel** [comment]
End of review round. The code IS much more readable, now. Thanks.

**Comment by jxchinaren** [comment]
@paulmartel Thanks for reviewing that, I added another commit to resolve your comments.

**Pull request merged**